feat(components): Camel MDC

squakez opened the pull request:
# Description
Closes CAMEL-21889 which implements proposal in https://github.com/apache/camel/blob/main/proposals/mdc.adoc

With this in place, the user will have just to provide the component in the pom and additionally the `camel.mdc.enabled=true` (plus eventually other configurations as shown in doc). Then he can show the MDC information via regular logging variables (ie, in log4j, `%X{exchangeID}`).

- [x] Core components architecture
- [x] Include headers
- [x] Include user headers
- [x] Unit test
- [x] Doc / Proposal upgrade

I will develop separately the following tasks:

- [ ] CSB starter
- [ ] Quarkus extension
- [ ] Previous MDC deprecation (likely in 4.16)

<!--
- Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
-->

# Target

- [x] I checked that the commit is targeting the correct branch (Camel 4 uses the `main` branch)

# Tracking
- [x] If this is a large change, bug fix, or code improvement, I checked there is a [JIRA issue](https://issues.apache.org/jira/browse/CAMEL) filed for the change (usually before you start working on it).

<!--
# *Note*: trivial changes like, typos, minor documentation fixes and other small items do not require a JIRA issue. In this case your pull request should address just this issue, without pulling in other changes.
-->

# Apache Camel coding standards and style

- [x] I checked that each commit in the pull request has a meaningful subject line and body.

<!--
If you're unsure, you can format the pull request title like `[CAMEL-XXX] Fixes bug in camel-file component`, where you replace `CAMEL-XXX` with the appropriate JIRA issue.
-->

- [x] I have run `mvn clean install -DskipTests` locally from root folder and I have committed all auto-generated changes.

<!--
You can run the aforementioned command in your module so that the build auto-formats your code. This will also be verified as part of the checks and your PR may be rejected if if there are uncommited changes after running `mvn clean install -DskipTests`.

You can learn more about the contribution guidelines at https://github.com/apache/camel/blob/main/CONTRIBUTING.md
-->

davsclaus commented on components/camel-mdc/pom.xml:
```diff
@@ -0,0 +1,63 @@
+  <artifactId>camel-mdc</artifactId>
+  <packaging>jar</packaging>
+  <name>Camel :: MDC</name>
+  <description>Mapped Diagnostic Context component</description>
```
I dont think many knows this is related to logging. So maybe

Logging MDC support

davsclaus commented on components/camel-mdc/src/main/java/org/apache/camel/mdc/MDCProcessorsInterceptStrategy.java:
```diff
@@ -0,0 +1,64 @@
+        return new DelegateAsyncProcessor(new TraceProcessor(target));
+    }
+
+    private class TraceProcessor implements Processor {
```
This would force the downstream processor to be processed synchronously - That works with MDC because it would be the same thread that continue processing .... 

The async is hard to get correct with MDC and we have MDCUnitOfWork doing this, and then some special code in the routing engine when Camel routing engine goes async.

davsclaus commented on core/camel-main/src/main/docs/main.adoc:
```diff
@@ -562,6 +562,15 @@ The camel.telemetryDev supports 4 options, which are listed below.
 
 
+=== Camel MDC configurations
+The camel.mdc supports 0 options, which are listed below.
```
Oh something seems wrong here saying 0 options

davsclaus commented on core/camel-main/src/main/java/org/apache/camel/main/MdcConfigurationProperties.java:
```diff
@@ -0,0 +1,57 @@
+ * Global configuration for MDC
+ */
+@Configurer(extended = true)
+public class MdcConfigurationProperties implements BootstrapCloseable {
```
Looks like there should be aded some options here you can configure

davsclaus reviewed (commented):
See something like AsyncEndpointRecipientList2Test that has a special test component that process async, which makes Camel continue routing in another thread. This can maybe be used for MDC unit test to see if all this still work when this happens.

squakez commented:
I've fixed the problems, it should be good now. I have also changed the test and included a thread verification name on the async execution to make sure the async component (wiretap) is really executing on a different thread and still the MDC is carried over.

In order to clarify the intent of this PR I've added some more documentation that I hope it clarifies the main aspects of why we should move from the older MDC to this newer implementation. Here an extract:

> The goal of this component is to avoid to work on low level API in Java. In older MDC implementations you had to hack into the code to include MDC such as:

```java
    org.slf4j.MDC.put("myCustomMDCKey", "myCustomKeyValue");
```
> And later you had to make sure to provide MDC context propagation in async components (eg, `wiretap`) in order to make sure to have such context available in the new executing async thread. With this new service, the only thing to do is to add the value as a Camel Exchange header (or a property), for example:
```java
    .setHeader("myCustomMDCKey", simple("myCustomKeyValue"))
```
> include the MDC service and additionally instruct the Camel application to treat that header as a MDC trace (via `camel.mdc.customHeaders=myCustomMDCKey` or `*` to include all headers). You won't need any longer to worry about context propagation as the propagation will be done via Camel Exchange instead.
>
>NOTE: you won't also need any longer to hack the code using Java DSL, as you can put headers in any Camel DSL.
>
>Depending on the logging technology used, you can now include the MDC parameters you want to trace in your logging configuration. For example, in `log4j2` configuration you can include them as shown below:
>
```
... [%X{camel.contextId}, %X{camel.routeId}, %X{camel.exchangeId}, %X{camel.messageId}, %X{customHead}, %X{customProp}]
```
>
>During the execution you can verify the output of the log to see the traces appended to your logger.

Basically we foster a higher level of abstraction to let the user have the feature without the need to do things at a low level. We also make this available to any other DSL, which was not possible before. We simplify also the long term maintenance as we have higher cohesion in a unique component instead of code dealing in multiple places. And we consistently have a context propagation, leveraging the Exchange mechanism instead of any other lower level threading mechanism.

Feel free to ask for any further clarification. Thanks!

davsclaus commented on core/camel-main/src/main/java/org/apache/camel/main/BaseMainSupport.java:
```diff
@@ -1851,6 +1869,28 @@ private void setOtel2Properties(
+        boolean enabled = obj != null ? CamelContextHelper.parseBoolean(camelContext, obj.toString()) : true;
+        if (enabled) {
+            CamelMDCService mdc = resolveMDCService(camelContext);
+            setPropertiesOnTarget(camelContext, mdc, mdcProperties, "camel.mdc.", failIfNotSet, true,
```
Ah okay, yeah so that was for ignore case

squakez commented on core/camel-main/src/main/java/org/apache/camel/main/MdcConfigurationProperties.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    /**
+     * Provide the headers you would like to use in the logging. Use `*` value to include all available headers
```
It seems the way used in other properties is the `&#42;` character. I've used that one.

squakez commented on core/camel-main/src/main/java/org/apache/camel/main/MdcConfigurationProperties.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    /**
+     * Provide the properties you would like to use in the logging. Use `*` value to include all available properties
```
Good one. Yes, I've changed to exchangeProperties and exchangeHeaders as well.

squakez commented:
> Can you add unit test with that async component I pointed previously. Wiretap and Async Component does not execute fully similar.

Okey, I've added the test and I had to refactor to manage an AsyncProcessor. Now it seems more consistent and the new thread spanned from the async processors are properly sharing MDC context as well.
